src/Lumi/Components2/Clip.purs

```diff
@@ -0,0 +1,112 @@
+import Effect.Class (liftEffect)
+import Effect.Console as Console
+import Effect.Uncurried (EffectFn1, EffectFn3, mkEffectFn1, runEffectFn3)
+import Effect.Unsafe (unsafePerformEffect)
+import Lumi.Components (LumiComponent, lumiComponent, lumiElement)
+import Lumi.Components.Spacing (Space(..))
+import Lumi.Components2.Box (box)
+import Lumi.Components2.Button (_linkStyle, button)
+import Lumi.Styles (styleModifier_, toCSS)
+import Lumi.Styles.Box (FlexAlign(..), _justify)
+import Lumi.Styles.Box as Styles.Box
+import Lumi.Styles.Clip as Styles.Clip
+import Lumi.Styles.Theme (LumiTheme(..), useTheme)
+import React.Basic.DOM as R
+import React.Basic.Emotion as E
+import React.Basic.Hooks (Hook, JSX, Ref, UseState, coerceHook, readRefMaybe, useRef, useState, (/\), type (/\))
+import React.Basic.Hooks as React
+import React.Basic.Hooks.Aff (UseAff, useAff)
+import React.Basic.Hooks.ResetToken (ResetToken, UseResetToken, useResetToken)
+import Web.DOM (Node)
+
+type ClipProps
+  = ( content :: Array JSX
+    )
+
+clip :: LumiComponent ClipProps
+clip =
+  unsafePerformEffect do
+    lumiComponent "Clip" defaults \props -> React.do
+      theme@(LumiTheme { colors }) <- useTheme
+      ref <- useRef Nullable.null
+      { copied, copy } <- useClip ref
+      let
+        copyButton =
+          lumiElement button
+            $ _linkStyle
+            $ styleModifier_
+                ( E.merge
+                    [ E.css
+                        { marginLeft: E.prop S16
+                        , lineHeight: E.str "12px"
+                        }
+                    , guard copied do
+                        E.css
+                          { color: E.color colors.black1
+                          , "&:hover": E.nested $ E.css { textDecoration: E.none }
+                          }
+                    ]
+                )
+            $ _
+                { content = [ R.text if copied then "Copied!" else "Copy" ]
+                , onPress = copy
+                }
+      pure
+        $ E.element R.div'
+            { className: props.className
+            , css: toCSS theme props Styles.Clip.clip
+            , children:
+              [ E.element R.div'
+                  { className: ""
+                  , css:
+                    toCSS theme props
+                      $ Styles.Box.box
+                      >>> Styles.Box._justify Center
+                  , ref
+                  , children: props.content
+                  }
+              , lumiElement box
+                  $ _justify Center
+                  $ _ { content = [ copyButton ] }
+              ]
+            }
+  where
+  defaults = { content: [] }
+
+newtype UseClip hooks
+  = UseClip (UseAff (ResetToken /\ Boolean) Unit (UseState Boolean (UseResetToken hooks)))
+
+derive instance ntUseClip :: Newtype (UseClip hooks) _
+
+useClip :: Ref (Nullable Node) -> Hook UseClip { copied :: Boolean, copy :: Effect Unit }
```

Comment:
I wonder if we should have `copied` be a three-valued enum so that we can properly communicate failures in the component?

Reply:
Hmm, that might make sense. Curious what you think after seeing my other response above about highlighting.

Reply:
Yes, that makes sense now. Thanks! :)